cardinal/plugin_systemdelayer.go

```diff
@@ -0,0 +1,112 @@
+package cardinal
+
+import (
+	"github.com/rotisserie/eris"
+)
+
+type SystemDelayManager interface {
+	AddTask(func(ctx WorldContext) error, int) error
+	AmountOfTasks() int
+	ClearTasks()
+	Register(*World) error
+}
+
+// task represents a task that can be added to a task queue.
+type task struct {
+	delay int
+	task  func(ctx WorldContext) error
+}
+
+// SystemDelayer represents a type that manages a queue of tasks to be executed with a specified delay.
+type SystemDelayer struct {
+	tasks []*task
+}
+
+// newSystemDelayer returns a new instance of SystemDelayer.
+func newSystemDelayer() *SystemDelayer {
+	return &SystemDelayer{
+		tasks: make([]*task, 0),
+	}
+}
+
+// AddTask adds a new task to the task queue with the given system and delay.
+// The delay is decremented by one, and if it becomes less than zero, an error is returned.
+// The task is appended to the tasks slice with the specified delay and system.
+// Parameters:
+// - system: a cardinal.System object representing the system to be executed as a task.
+// - delay: an unsigned integer representing the delay in seconds before the task is executed.
+// Returns:
+// - error: an error object if the delay is less than zero, nil otherwise.
+func (s *SystemDelayer) AddTask(taskFunc func(ctx WorldContext) error, delay int) error {
+	delay--
```

Comment:
From a very high level perspective, this method is to allow the user to add a function to be executed N ticks later from now. 

The logic is executed from wCtx.DelatedTask(...) at a high level, then passed down to this method. The logic is executed later in a system.

Reply:
This comment was specifically referring to the delay-- line [](edited-with-Axolo)

Reply:
It's to match user intuition. If you put in 1 for the delay without the decrement it will actually will not execute in the next tick as you expect. In the next tick the delay will decrement to zero, and in the next next tick it will finally execute the task. I'll add a comment here for clarity.

Reply:
It seems like you are making an assumption that each tick is 1 second? This is not necessarily the case since the tick rate is configurable.

If you are actually delaying based on time, then the delay should also be on millisecond precision since that is the granularity that we support for the ticks. Additionally, if you are delaying based on time, the only thing you would have to do is track `current_tick_timestamp_in_ms + delay_in_ms` then you can just trigger the task when that value is less than the current tick timestamp in the future. There is no need to manually decrement this. [](edited-with-Axolo)

Reply:
no delay is based on tick not time, that was my intent? I can do delay based on time if you prefer [](edited-with-Axolo)

Reply:
Your comment says: `- delay: an unsigned integer representing the delay in seconds before the task is executed.` [](edited-with-Axolo)

Reply:
oh that's an error, sorry it was generated by AI [](edited-with-Axolo)

Reply:
I would imagine the user would prefer to delay based on time. Regardless, I don't think the way you should be tracking the delay is by decrementing the numbers at every tick; you should just diff the delta. [](edited-with-Axolo)

Reply:
I'll update it to do time [](edited-with-Axolo)

Reply:
and the delta [](edited-with-Axolo)

Reply:
wait, for determinism I think it should store remaining time and decrement. If we store the timestamp things can go off.

I can decrement the time based off of ms per tick. [](edited-with-Axolo)

Reply:
because if the task was designed to trigger 2 ticks later (based on ms) during a crash..... then you recover 2 days later, the tasks will execute instantly. You still want the tasks to execute 2 ticks later after recovery. [](edited-with-Axolo)

Reply:
That's not what it means to be deterministic in Cardinal.

Cardinal simply expects that after the tick is finalized (and the tick timestamp + transaction data is written via the router to the data availability layer). A new cardinal instance can rebuild the state from tick 0 by replaying the recovered timestamps and transaction data.

In your example, this task was never executed in the first place, but when it does get executed immediately on the next tick (instead of 2 * tick interval) as you expected, then if you replay that tick then it will result in the same state. [](edited-with-Axolo)

Reply:
[Slack File: CleanShot 2024-07-25 at 01.52.15@2x.png](https://files.slack.com/files-pri/T03G06L71RD-F07DGKF6PHD/cleanshot_2024-07-25_at_01.52.15_2x.png) [](edited-with-Axolo)

Reply:
[Slack File: CleanShot 2024-07-25 at 01.52.45@2x.png](https://files.slack.com/files-pri/T03G06L71RD-F07DX381EG3/cleanshot_2024-07-25_at_01.52.45_2x.png) [](edited-with-Axolo)

Reply:
Let's assume 1 sec per tick.

The user api is like sleep(5) so the instruction from the user expects is to execute task 5 seconds (or 5 ticks)  from now. The user sort of expects this.

So in this case for the example if we recover at tick 0 from a crash, then we tick forward the timestamp jump will be huge if ran far into the future causing the task to execute immediately at tick 1 when the user expected it to be 5 ticks from tick zero. [](edited-with-Axolo)

Reply:
That's a non-issue; that would be expected behavior when the task delay is based on timestamp. If I queue a task that says "Construct building A after 30 minutes", then I expect the building to be constructed after 30 minutes regardless of whether or not cardinal is ticking during those 30 minutes period. [](edited-with-Axolo)

Reply:
You can't use ticks and time interchangeably; it's not the same measure of time. [](edited-with-Axolo)

Reply:
I think it's better to time it relative to ticks and ticks per second. So users can say 5 seconds after tick 0. Which means at 1 tick/s the user request is translated to execute on tick 5.

The reason is for example let's say a player dies at tick 0 and he's supposed to drop coins 5 seconds later.

If we replay at tick 1 in the far future those coins will be there. But in the game itself if it didn't crash those coins aren't supposed to exist yet.

Or another more clearer example we use this DelayTask to trigger the game round to end 1 hour from game start (1 hour after tick 0).

if we crash right after tick 0 and we recover and run the game 1 day later.... the game will immediately end at tick 1 because it's well past 1 hour. The user however expects the game to end 1 hour after game start. [](edited-with-Axolo)

Reply:
I think the gut intuition a user feels when he/she sees this:

```
wCtx.DelayTask(f, 5)
```
Is that the task `f`  will execute 5s from the current tick. Not 5s from the current timestamp. So executing the tick at different times results in different execution times, but same execution at tick numbers. [](edited-with-Axolo)

Reply:
>  if we crash right after tick 0 and we recover and run the game 1 day later.... the game will immediately end at tick 1 because it's well past 1 hour. The user however expects the game to end 1 hour after game start.

If it doesn't end at tick 1, when does it end? If you measure from the actual match start, it is well past 1 hour already, so it's intended behavior for the match to have already finished.

There is also no reliable way to translate seconds <> tick, because in reality each tick might take longer than what it's expected to due to tick drift caused by ticks taking longer than expected to complete.


Again, you can't be translating tick <> timestamp interchangeably. If you want to delay by wall-clock time, then your task will be executed when the wall-clock time has elapsed.


If you want to delay task execution by actual discrete ticks, then the passed in parameter should be ticks instead of time and could be a separate API/method entirely. [](edited-with-Axolo)

Reply:
Delay by time:
```
if (wCtx.Timestamp >= time_to_execute) {

 // execute task
}
```
Delay by tick (the behavior you are describing):
```
if (wCtx.CurrentTick >= tick_to_execute) {

  // execute task
}
```
These are 2 different behaviors, which might have its own different usecases, but my point here is that

1) this is not invalid non-deterministic behavior; these are equally valid, but different behavior.
2) if you want to delay based on discrete tick steps (cardinal is running), then you should delay by X ticks instead of "delay by X wall-clock millisecond, implicitly converted to ticks" [](edited-with-Axolo)

Reply:
>> when does it end?
> The dev expects the game to end 1 hour from tick 0. If ticks execute at 1tick/s then this means 60*60 = tick 3600 it should end.
>> There's no reliable way to translate seconds
> If we can't reliably translate tick speeds then using time to determine it will always be non-deterministic if we use time as for determinism we need future events happening on the exact same tick.
>> You can't be translating tick <> timestamp
> You're right we can't. But this is only because we can't reliably have ticks execute at a certain rate.
>> If you want to delay task execution by actual discrete ticks,
> That's how it's currently set up now. Delay by ticks. I think it's better to leave it this way now given what we've just talked about.
>> These are 2 different behaviors, which might have its own different usecases, but my point here is that...
> Yeah I was originally doing delay by ticks. Let's just stick with that. I think for my examples it works better. It will be not deterministic if I use time relative to anything else other then a timestamp due to variability of tick rate. [](edited-with-Axolo)

Reply:
I'm just going to stick with ticks instead of time if you're ok with that [](edited-with-Axolo)

Reply:
This is not a non-determinism issue. If you don't implement the one with time, I'll just implement it myself [](edited-with-Axolo)

Reply:
I can do both how about that? [](edited-with-Axolo)

Reply:
Up to you [](edited-with-Axolo)